spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultTaskService.java

```diff
@@ -76,90 +85,85 @@
 		this.dataSourceProperties = dataSourceProperties;
-		this.repository = repository;
-		this.deploymentIdRepository = deploymentIdRepository;
+		this.taskDefinitionRepository = taskDefinitionRepository;
+		this.taskExecutionRepository = taskExecutionRepository;
+		this.taskExplorer = taskExplorer;
 		this.registry = registry;
 		this.taskLauncher = taskLauncher;
 		this.resourceLoader = resourceLoader;
 		this.whitelistProperties = new WhitelistProperties(metaDataResolver);
 	}
 
-	private TaskDefinition updateTaskProperties(TaskDefinition taskDefinition) {
-		TaskDefinitionBuilder builder = TaskDefinitionBuilder.from(taskDefinition);
-		builder.setProperty("spring.datasource.url", dataSourceProperties.getUrl());
-		builder.setProperty("spring.datasource.username",
-				dataSourceProperties.getUsername());
-		// password may be empty
-		if (StringUtils.hasText(dataSourceProperties.getPassword())) {
-			builder.setProperty("spring.datasource.password",
-					dataSourceProperties.getPassword());
-		}
-		builder.setProperty("spring.datasource.driverClassName",
-				dataSourceProperties.getDriverClassName());
-		return builder.build();
-	}
-
 	@Override
 	public void executeTask(String taskName, Map<String, String> taskDeploymentProperties,
 			List<String> commandLineArgs) {
 		Assert.hasText(taskName, "The provided taskName must not be null or empty.");
 		Assert.notNull(taskDeploymentProperties,
 				"The provided runtimeProperties must not be null.");
-		TaskDefinition taskDefinition = this.repository.findOne(taskName);
+		TaskDefinition taskDefinition = this.taskDefinitionRepository.findOne(taskName);
 		if (taskDefinition == null) {
 			throw new NoSuchTaskDefinitionException(taskName);
 		}
 
-		taskDefinition = this.updateTaskProperties(taskDefinition);
-
 		URI uri = this.registry.find(String.format("task.%s", taskDefinition.getRegisteredAppName()));
 		Resource resource = this.resourceLoader.getResource(uri.toString());
-
+		TaskExecution taskExecution = taskExecutionRepository.createTaskExecution();
+		taskDefinition = this.updateTaskProperties(taskDefinition, taskExecution);
 
 		Map<String, String> appDeploymentProperties = extractAppProperties(taskDefinition.getRegisteredAppName(), taskDeploymentProperties);
 		Map<String, String> deployerDeploymentProperties = DeploymentPropertiesUtils.extractAndQualifyDeployerProperties(taskDeploymentProperties, taskDefinition.getRegisteredAppName());
 		AppDefinition revisedDefinition = mergeAndExpandAppProperties(taskDefinition, resource, appDeploymentProperties);
+
 		AppDeploymentRequest request = new AppDeploymentRequest(revisedDefinition, resource, deployerDeploymentProperties, commandLineArgs);
 
 		String id = this.taskLauncher.launch(request);
 		if (!StringUtils.hasText(id)) {
 			throw new IllegalStateException("Deployment ID is null for the task:"
 					+ taskName);
 		}
-		String deploymentKey = DeploymentKey.forTaskDefinition(taskDefinition);
```

Comment:
Overlook from my side, especially 2).

The rationale, which we arrived at with @mminella is that for one given task, there may be several executions. The use of `DeploymentKey.forTaskDefinition()` as it were was hence completely flawed.

I suggest to do the minimum in this PR to avoid major breakage and open a second, separate PR, to get rid of `DeploymentKey.forTaskDefinition()`